mlflow/gateway/providers/mlflow.py

```diff
@@ -0,0 +1,143 @@
+        self.mlflow_config: MlflowModelServingConfig = config.model.config
+        self.headers = {"Content-Type": "application/json"}
+
+    @staticmethod
+    def process_payload(payload, key):
+        payload = jsonable_encoder(payload, exclude_none=True)
+
+        input_data = payload.pop(key)
+        request_payload = {"inputs": input_data if isinstance(input_data, list) else [input_data]}
+
+        if payload:
+            request_payload["params"] = payload
+
+        return request_payload
+
+    @staticmethod
+    def process_response(response_text):
+        return (
+            response_text if isinstance(response_text, (str, list)) else json.dumps(response_text)
+        )
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        # Example request to MLflow REST API server for completions:
+        # {
+        #     "inputs": ["hi", "hello", "bye"],
+        #     "params": {
+        #         "temperature": 0.5,
+        #         "top_k": 3,
+        #     }
+        # }
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_server_url,
+            path="invocations",
+            payload=self.process_payload(payload, "prompt"),
+        )
+
+        # Example response:
+        # {"predictions": ["hello", "hi", "goodbye"]}
+
+        return completions.ResponsePayload(
+            **{
+                "candidates": self.process_response(resp["predictions"]),
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        # Example request to MLflow REST API for chat:
+        # {
+        #     "inputs": ["question"],
+        #     "params": ["temperature": 0.2],
+        # }
+
+        payload = self.process_payload(payload, "messages")
+
+        query_count = len(payload["inputs"])
+        if query_count > 1:
+            raise HTTPException(
+                status_code=422,
+                detail="MLflow chat models are only capable of processing a single query at a "
+                f"time. The request submitted consists of {query_count} queries.",
+            )
+
+        payload["inputs"] = [payload["inputs"][0]["content"]]
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.mlflow_server_url,
+            path="invocations",
+            payload=payload,
+        )
+
+        # Example response:
+        # {"predictions": "answer"}
+
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {
+                            "role": "assistant",
+                            "content": self.process_response(resp["predictions"]),
```

Comment:
If `candidate_count` is > 1, then the response will be a list of message completions, so embedding them directly in the `"content"` section of a single message won't be correct.

I also think it's somewhat risky to assume that the model is only going to return a single column of text data. If a dataframe is returned, the response format breaks down as well.

As with completions, can we assume that the format for the dict contained under `"predictions"` is 

```
{"candidates": ["chat completion 1", "chat completion 2", ...]}
```

and validate against it?

Reply:
If only a single string column is defined, using it seems fine too, even if the column name isn't `candidates`. This will help with transformers compatibility until we have standardized signatures for completions in pyfunc. Regardless, we should valid the structure of `resp["predictions"]`.